timex/timex_test.go

```diff
@@ -0,0 +1,45 @@
+package timex
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"golang.org/x/sync/errgroup"
+)
+
+func TestTicker(t *testing.T) {
+	ticker := NewTicker(2 * time.Millisecond)
+	time.AfterFunc(5*time.Millisecond, func() {
+		ticker.Stop()
+	})
+	cnt := 0
+	for range ticker.C {
+		cnt++
+	}
+	require.Equal(t, 2, cnt)
```

Comment:
this test might be flaky under heavy load as a 1ms delay on the AfterFunc may mean that the ticker fires 3 times.

Reply:
Really? I thought i t would it would tick at millisecond: 0, 2, 4 whereas stop would be fired at 5?

Reply:
From `time.NewTimer()`:
> NewTimer creates a new Timer that will send the current time on its channel after at least duration d

Due to OS scheduling latency, all the timer can guarantee is that it will wait *at least* _n_ milliseconds before firing. It may take more if the process cannot be scheduled. Consider a cloudbuild machine with the standard machine type (https://cloud.google.com/cloud-build/pricing) - one vcpu. This is single threaded. If you have multiple concurrent build steps and one of them is this test, there could easily be multiple-millisecond scheduling latency. This means for example, the second tick of the ticker may not fire at 4ms because the process is not scheduled. it could be later. At that time, both the ticker and the afterfunc are due to fire and the afterfunc could be executed first. At that point, the count will be 1, and the ticker is stopped.

Reply:
So what do you suggest should I do? Not test? Test cnt >= 1 and <= 4 (test hasn't dialed yet btw. Initially I had microseconds and u saw the variance more starkly, but I know that's not a precise measure).

Reply:
I dont know. This is an area that needs better testing support - using a fake clock to simulate real time so 1) the test is accurate in the presence of scheduling latency, and 2) you do not actually have to insert delays into test code, which just slows them down.

I've had the same problem before, but also with a timeline that related to rendezvous on channels and ordering. I don't know yet how to test this sort of stuff.

Reply:
If this is fine, then perhaps leave it, but add a comment that it may perhaps be flaky.

Reply:
I'm thinking now the ticks I get are: 2,4, closeChannel tick. 
However, if I run it as: 
```
	for _, ok := <-ticker.C; ok; _, ok = <-ticker.C {
		cnt++
	}
```
I still get the same result (would have expected cnt being 1)

Reply:
the close channel tick had `ok == false` so you never entered the loop for that one.